Tenant-Aware Report filters and ValidatedSW. (#574)
nautobot/nautobot-plugin-device-lifecycle-mgmt: docs/user/software_lifecycle.md

Validated Software object can be assigned to:

- devices
- device tenants *(honored only when `multi_tenant_mode=True` — see [Multi-tenant mode](#multi-tenant-mode-opt-in))*
- device types
- device roles
- inventory items
- object tags applied to devices and inventory items

One Validated Software object can be assigned to multiple other objects.

> ## Multi-tenant mode (opt-in)

Reviewer: This probably belongs under the `## Validated Software matching logic` heading below?

- For tenanted devices in multi-tenant mode, direct `devices` and `object_tags` matches are tenant-agnostic and can match VS records with any (or no) tenant assigned.

### For devices — legacy mode (`multi_tenant_mode=False`)

Validated Software will be used if one, or more, of the following applies:

- Device is explicitly listed in the Validated Software `devices` attribute.
- Device's device type AND device role match `device_types` AND `device_roles` in Validated Software. This applies only if BOTH are set. See the **Special cases** subsection that follows.
- Device's device type is listed in the Validated Software `device_types` attribute.
- Device's role is listed in the Validated Software `device_roles` attribute.
- Device's tags are listed in the Validated Software `object_tags` attribute.

Only ValidatedSoftware records with no `device_tenants` set are considered. The device's own tenant has no effect on matching.

### For devices — multi-tenant mode (`multi_tenant_mode=True`)

If the device has a tenant assigned, Validated Software will be used if one, or more, of the following applies:

> 
> - Device is explicitly listed in the Validated Software `devices` attribute.
> - Device's tenant AND device type match `device_tenants` AND `device_types` in Validated Software. This applies only if BOTH are set. See the **Special cases** subsection that follows.
> - Device's tenant AND device role match `device_tenants` AND `device_roles` in Validated Software.
> - Device's tenant is listed in the Validated Software `device_tenants` attribute (with no device type set).
> - Device's tags are listed in the Validated Software `object_tags` attribute.

Reviewer 2: ```suggestion
- Device is explicitly listed in the Validated Software `devices` attribute.
- The record's `device_tenants` includes the device's tenant, AND both `device_types` and `device_roles` are set and match the device. See the **Special cases** subsection that follows.
- Device's tenant matches a `device_tenants` entry AND `device_types` matches the device's type (when `device_types` is set but `device_roles` is not set on the record).
- Device's tenant matches a `device_tenants` entry AND `device_roles` matches the device's role (when `device_roles` is set but `device_types` is not set on the record).
- Device's tenant matches a `device_tenants` entry and neither `device_types` nor `device_roles` is set on the record (tenant-only record).
- Device's tags are listed in the Validated Software `object_tags` attribute.
```

When resolving whether Validated Software is taken into account when validating software on a given device, the following logic applies.

### Matching workflow for devices

The following diagram summarizes how a candidate Validated Software record is accepted or rejected for a given device under each mode. The bullet-list subsections below expand on the individual match criteria referenced in the diagram.

```mermaid
flowchart TD
    Start([Device + Validated Software candidate]) --> Flag{multi_tenant_mode?}

>     Flag -- "False / legacy" --> LegacyVSTenant{VS has<br/>device_tenants set?}

Reviewer: I think we should update `VS` throughout this diagram to `ValidatedSoftware`

>     LegacyVSTenant -- Yes --> Exclude([Exclude])

Reviewer: ```suggestion
    LegacyVSTenant -- Yes --> Exclude([Device and Validated Software do not match])
```

    LegacyVSTenant -- No --> LegacyMatch{Match on devices,<br/>device_types,<br/>device_roles,<br/>or object_tags?}

>     LegacyMatch -- Yes --> Include([Include in queryset])

Reviewer: ```suggestion
    LegacyMatch -- Yes --> Include([Device and Validated Software match])
```
